cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go

```diff
@@ -106,49 +165,36 @@ func ResetClusterStatusForNode(nodeName string, client clientset.Interface) erro
 func UploadConfiguration(cfg *kubeadmapi.InitConfiguration, client clientset.Interface) error {
 	fmt.Printf("[upload-config] storing the configuration used in ConfigMap %q in the %q Namespace\n", kubeadmconstants.KubeadmConfigConfigMap, metav1.NamespaceSystem)
 
-	// Prepare the ClusterConfiguration for upload
-	// The components store their config in their own ConfigMaps, then reset the .ComponentConfig struct;
-	// We don't want to mutate the cfg itself, so create a copy of it using .DeepCopy of it first
-	clusterConfigurationToUpload := cfg.ClusterConfiguration.DeepCopy()
-	clusterConfigurationToUpload.ComponentConfigs = kubeadmapi.ComponentConfigs{}
-
-	// Marshal the ClusterConfiguration into YAML
-	clusterConfigurationYaml, err := configutil.MarshalKubeadmConfigObject(clusterConfigurationToUpload)
+	lock, err := resourcelock.New(resourcelock.ConfigMapsResourceLock, metav1.NamespaceSystem, kubeadmconstants.KubeadmConfigLockConfigMap, client.CoreV1(), nil, resourcelock.ResourceLockConfig{
+		Identity: kubeadmconstants.KubeadmConfigLockConfigMap,
+	})
 	if err != nil {
-		return err
+		return errors.Wrap(err, "failed to create a configmap resource lock")
 	}
 
-	// Prepare the ClusterStatus for upload
-	// Gets the current cluster status
-	// TODO: use configmap locks on this object on the get before the update.
-	clusterStatus, err := configutil.GetClusterStatus(client)
+	ctx, cancel := context.WithCancel(context.TODO())
+	res := make(chan error, 1)
+
+	le, err := leaderelection.NewLeaderElector(leaderelection.LeaderElectionConfig{
```

Comment:
i wonder if you can maintain a single LeaderElector and a single ConfigMapLock for all methods that modify the ConfigMap, or at least have a helper function that creates them for us. 
after that we can update the OnStartedLeading callbacks to comply with the different use cases.

Reply:
I'll check tomorrow :+1: